src/modules/bulletproofs/inner_product_impl.h

```diff
@@ -0,0 +1,848 @@
+ * from the big comment block above, are chosen; and (b) when/how each `x_k^-2` is computed.
+ */
+static int secp256k1_bulletproof_innerproduct_vfy_ecmult_callback(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *data) {
+    secp256k1_bulletproof_innerproduct_vfy_ecmult_context *ctx = (secp256k1_bulletproof_innerproduct_vfy_ecmult_context *) data;
+
+    /* First 2N points use the standard Gi, Hi generators, and the scalars can be aggregated across proofs.
+     * Inside this if clause, `idx` corresponds to the index `i` in the big comment, and runs from 0 to `2n-1`.
+     * Also `ctx->vec_len` corresponds to `n`. */
+    if (idx < 2 * ctx->vec_len) {
+        /* Number of `a` scalars in the proof (same as number of `b` scalars in the proof). Will
+         * be 2 except for very small proofs that have fewer than 2 scalars as input. */
+        const size_t grouping = ctx->vec_len < IP_AB_SCALARS / 2 ? ctx->vec_len : IP_AB_SCALARS / 2;
+        const size_t lg_grouping = secp256k1_floor_lg(grouping);
+        size_t i;
+        VERIFY_CHECK(lg_grouping == 0 || lg_grouping == 1); /* TODO support higher IP_AB_SCALARS */
+
+        /* Determine whether we're multiplying by `G_i`s or `H_i`s. */
+        if (idx < ctx->vec_len) {
+            *pt = ctx->geng[idx];
+        } else {
+            *pt = ctx->genh[idx - ctx->vec_len];
+        }
+
+        secp256k1_scalar_clear(sc);
+        /* Loop over all the different inner product proofs we might be doing at once. Since they
+         * share generators `G_i` and `H_i`, we compute all of their scalars at once and add them.
+         * For each proof we start with the "seed value" `ctx->proof[i].xcache[0]` (see next comment
+         * for its meaning) from which every other scalar derived. We expect the caller to have
+         * randomized this to ensure that this wanton addition cannot enable cancellation attacks.
+         */
+        for (i = 0; i < ctx->n_proofs; i++) {
+            /* To recall from the introductory comment: most `s_i` values are computed by taking an
+             * earlier `s_j` value and multiplying it by some `x_k^2`.
+             *
+             * We now explain the index `j`: it is the largest number with one fewer 1-bits than `i`.
+             * Alternately, the most recently returned `s_j` where `j` has one fewer 1-bits than `i`.
+             *
+             * To ensure that `s_j` is available when we need it, on each iteration we define the
+             * variable `cache_idx` which simply counts the 1-bits in `i`; before returning `s_i`
+             * we store it in `ctx->proof[i].xcache[cache_idx]`. Then later, when we want "most
+             * recently returned `s_j` with one fewer 1-bits than `i`, it'll be sitting right
+             * there in `ctx->proof[i].xcache[cache_idx - 1]`.
+             *
+             * Note that `ctx->proof[i].xcache[0]` will always equal `-a_1 * prod_{i=1}^{n-1} x_i^-2`,
```

Comment:
Could maybe be checked in DEBUG/TEST mode.

Reply:
I don't think so, because we don't know `-a_1` explicitly.